### PR TITLE
Migrate renorm kernels from sgl-kernel to FlashInfer JIT

### PR DESCRIPTION
**Pull request by Johnsonms**
Migrate top_k_renorm_probs, top_p_renorm_probs, and top_k_mask_logits from compiled sgl-kernel CUDA implementations to FlashInfer's JIT-compiled public API. This reduces the sgl-kernel wheel size and leverages FlashInfer's optimized and maintained implementations.

Changes:
- Use flashinfer.sampling public API instead of torch.ops.sgl_kernel
- Remove CUDA kernel bindings from common_extension.cc and common_extension_musa.cc
- Remove function declarations from sgl_kernel_ops.h
- Remove renorm.cu from build (keep sampling.cu for other functions)
- Add comprehensive benchmark suite (bench_renorm.py)

## Motivation
https://github.com/sgl-project/sglang/issues/17865
move (external) flashinfer/csrc/renorm.cu: https://github.com/flashinfer-ai/flashinfer/blob/bc29697ba20b7e6bdb728ded98f04788e16ee021/csrc/renorm.cu (≈4MB)

## Accuracy Tests
`python -m pytest sgl-kernel/tests/test_sampling.py -s`
image

```
python -m sglang.launch_server --model Qwen/Qwen3-4B
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1319 --parallel 128
```
image

```
python -m sglang.launch_server --model /data/Qwen3-8B/ --disable-radix-cache
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1319 --parallel 128
```
image

## Benchmarking and Profiling
`python benchmark/bench_renorm.py`

Correctness check
image

Benchmarking
1. Benchmarking top_k_renorm_probs...
image

2. Benchmarking top_p_renorm_probs...
image

3. Benchmarking top_k_mask_logits...
image

Used the data mades sumary as following:
image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
3. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
5. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
6. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: sgl-kernel/csrc/common_extension_musa.cc
Comment (DarkSharpness) [inline sgl-kernel/csrc/common_extension_musa.cc:1]: One question: is this change safe? I'm not familiar with musa. Will this change break musa support (e.g. flashinfer is not supported by musa platform.

Comment (Johnsonms) [reply sgl-kernel/csrc/common_extension_musa.cc:1]: Thanks @DarkSharpness , there is an exactly potential issue

I investigated the MUSA path and made an additional change to improve MUSA safety. Below are my conclusions.

The key point is the **platform dispatch pattern** introduced in the latest commit (`19f055084`). For each of the three renorm-related functions, the code explicitly checks `device.type` to select the correct backend.

**Dispatch behavior:**

**`top_k_renorm_probs` (lines 79–82)**

```python
if probs.device.type == "musa":
    return _top_k_renorm_probs_internal(...)   # uses torch.ops.sgl_kernel (precompiled C++)
else:
    return flashinfer.sampling.top_k_renorm_probs(...)  # FlashInfer JIT (CUDA only)
```

The same dispatch pattern is used in:

- **`top_p_renorm_probs` (lines 118–121)**
- **`top_k_mask_logits` (lines 463–466)**

**Summary:**

- **MUSA path (`device.type == "musa`)**
    
    Calls `torch.ops.sgl_kernel.*`, which are the precompiled C++ operators that were already working prior to this PR. FlashInfer is not involved.
    
- **CUDA path (`else`)**
    
    Calls `flashinfer.sampling.*`, using the new JIT-based CUDA implementation.
    

Additionally, `renorm.cu` must be retained for **MUSA compilation** in [here ](https://github.com/sgl-project/sglang/blob/34d975b18fa5a01b0950735faac85b79dc57507b/sgl-kernel/setup_musa.py#L81)(even though it is not listed in `CMakeLists.txt`), so that the `sgl_kernel` operators required by MUSA remain available.

With this setup, the existing MUSA behavior is preserved while enabling the new CUDA JIT path safely.

File: sgl-kernel/python/sgl_kernel/sampling.py [hunk 1]
```diff
@@ -452,7 +462,12 @@ def top_k_mask_logits(
     --------
     top_k_renorm_probs
     """
-    return _top_k_mask_logits_internal(logits, *_to_tensor_scalar_tuple(top_k))
+    if logits.device.type == "musa":
+        return _top_k_mask_logits_internal(logits, *_to_tensor_scalar_tuple(top_k))
+    else:
+        import flashinfer.sampling
```
Comment (BBuf) [inline sgl-kernel/python/sgl_kernel/sampling.py]: Can we move import to the top?

Comment (BBuf) [inline sgl-kernel/python/sgl_kernel/sampling.py]: Move import to the file top.

Comment (Johnsonms) [reply sgl-kernel/python/sgl_kernel/sampling.py]: Sure, done. Thanks @BBuf

Comment (Johnsonms) [reply sgl-kernel/python/sgl_kernel/sampling.py]: Sure, done. Thanks @BBuf

File: python/sglang/jit_kernel/benchmark/bench_renorm.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
+import itertools
+import os
+
+import sgl_kernel
+import torch
+import triton
+import triton.testing
+
+# CI environment detection
+IS_CI = (
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/benchmark/bench_renorm.py]: Please use this function to set `IS_CI`:
https://github.com/sgl-project/sglang/blob/0fd44ff342bb1e1ed7a110b5164034a288cc67a0/python/sglang/jit_kernel/benchmark/utils.py#L15-L20

File: python/sglang/jit_kernel/benchmark/bench_renorm.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+            k_val = min(max(top_k[i].item(), 1), vocab_size)
+            _, topk_indices = torch.topk(probs[i], k_val, largest=True)
+            mask = torch.zeros_like(probs[i])
+            mask[topk_indices] = 1.0
+            masked_probs = probs[i] * mask
+            renorm_probs[i] = masked_probs / (masked_probs.sum() + 1e-10)
+        return renorm_probs
+
+    # Vectorized renormalization
+    masked_probs = probs * mask
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/benchmark/bench_renorm.py]: Could this part of the code be moved inside the if statement? I think that would be clearer.

File: python/sglang/jit_kernel/benchmark/bench_renorm.py [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+            sorted_prob, indices = torch.sort(probs[i], descending=False)
+            cdf = torch.cumsum(sorted_prob, dim=-1)
+            mask = torch.zeros(vocab_size, dtype=torch.float32, device=probs.device)
+            mask.scatter_(0, indices, (cdf >= (1 - p_val) - eps).float())
+            masked_probs = probs[i] * mask
+            renorm_probs[i] = masked_probs / (masked_probs.sum() + eps)
+        return renorm_probs
+
+    # Vectorized renormalization
+    masked_probs = probs * mask
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/benchmark/bench_renorm.py]: Same comment as https://github.com/sgl-project/sglang/pull/18854/changes#r2857120570

**Review by HydraQYH — COMMENTED** [review]
Is this PR just one in a series of PRs? I haven't seen the migration code.

File: python/sglang/jit_kernel/benchmark/bench_renorm.py (hunk 2 above)
Comment (Johnsonms) [reply python/sglang/jit_kernel/benchmark/bench_renorm.py]: My mistake — I accidentally rebased it yesterday. I’ll fix it tomorrow. sorry @HydraQYH

File: python/sglang/jit_kernel/benchmark/bench_renorm.py (hunk 4 above)
Comment (Johnsonms) [reply python/sglang/jit_kernel/benchmark/bench_renorm.py]: Resolved, thanks

File: python/sglang/jit_kernel/benchmark/bench_renorm.py (hunk 3 above)
Comment (Johnsonms) [reply python/sglang/jit_kernel/benchmark/bench_renorm.py]: Great, resolved. Thanks

**Comment by Johnsonms** [comment]
> Is this PR just one in a series of PRs? I haven't seen the migration code.
Sorry, it is deleted by rebasing. right now it is restore. It's ready to review. Thanks @HydraQYH

File: python/sglang/jit_kernel/benchmark/bench_renorm.py (hunk 2 above)
Comment (HydraQYH) [reply python/sglang/jit_kernel/benchmark/bench_renorm.py]: Please use `is_in_ci`, I have no other comments besides this.

Comment (Johnsonms) [reply python/sglang/jit_kernel/benchmark/bench_renorm.py]: Resolve, Thanks!

**Review by Johnsonms — COMMENTED** [review]
Resolved

**Review by HydraQYH — APPROVED** [review]

**Review by BBuf — APPROVED** [review]

**Review by DarkSharpness — APPROVED** [review]
